File: trainer_app/Packages/TrainerCore/Sources/TrainerDomain/LegacySoapView.swift
import Foundation
import TrainerContracts

// Read-only view of a v1 (legacy) SOAP document (DF-009, AC-SOAP-06.2, V1-05 §5.6, PRD §11.4).
// Same rules as the Dart `LegacySoapView` (lib/models/legacy_soap_view.dart, DF-007).
//
// v1 documents have no `schemaVersion`. The view keeps every `structured.metrics` row and shows each one
// as 'uninterpretable' (해석 불가): no legacy row carries the condition fields a v2 metric needs
// (joint·motion·activeOrPassive, muscleGroup), so even a parsable number is not a catalog metric
// (V1-05 §5.6). Legacy notations (`ROM`, `통증`, `좌`, `도`, `작성 중`, ...) are kept as text and never
// mapped onto the contract enums, and values are never filled with 0 or with demo text.
//
// The view does not expose `diagnosis` (MIG-04) and has no write path: the stored document is private,
// and no public property returns `diagnosis` or an object that contains it.

/// One `structured.metrics[]` row of a v1 document, kept as stored.
public struct LegacyMetricView: Equatable, Sendable {
  /// The stored element. Normally an object with `type, label, side, value, unit, score, note`.
  public let source: JSONValue

  public init(_ source: JSONValue) {
    self.source = source
  }

  /// The stored row, or an empty object when the element is not an object.
  public var raw: [String: JSONValue] { source.objectValue ?? [:] }

  /// The legacy type as written: `rom`, `ROM`, `통증`, `test`, `general`, ...
  public var type: String? { raw["type"]?.stringValue }
  public var label: String? { raw["label"]?.stringValue }
  public var side: String? { raw["side"]?.stringValue }
  public var unit: String? { raw["unit"]?.stringValue }
  public var note: String? { raw["note"]?.stringValue }

  /// The value as stored (a string in every legacy writer), shown as text. Empty when absent or null.
  public var valueText: String {
    switch raw["value"] {
    case nil, .null?: return ""
    case let .string(text)?: return text
    case let .number(n)?: return n == n.rounded() && abs(n) < 1e15 ? String(Int64(n)) : String(n)
    case let .int(n)?: return String(n)
    case let .bool(b)?: return b ? "true" : "false"
    case let value?: return "\(value)"
    }
  }

  /// Always false: a legacy row is shown as '해석 불가' next to its original text (F-SOAP-06.2).
  public var isInterpretable: Bool { false }

  /// The label shown next to the original text.
  public var displayLabel: String { LegacySoapView.uninterpretableLabel }
}

/// Read-only view of a v1 SOAP document.
public struct LegacySoapView: Equatable, Sendable {
  /// Display label for legacy rows and unknown v2 metric codes.
  public static let uninterpretableLabel = "해석 불가"

  /// The document as read. Private: it holds `diagnosis`, which is never exposed (MIG-04).
  private let stored: [String: JSONValue]

  /// Every `structured.metrics` row, in stored order. Never shorter than the stored list.
  public let metrics: [LegacyMetricView]

  /// `structured.workflow.status` as written (`complete`, `완료`, `작성 중`, `공유됨`, ...).
  public let legacyStatus: String?

  /// `structured.workflow.completedCategories` as written (`S` or `subjective` style).
  public let completedCategories: [String]

  /// Reads a v1 document's `data` (the fixture tags already turned into `JSONValue` cases).
  public init(json: JSONValue) {
    stored = json.objectValue ?? [:]
    let structured = stored["structured"]
    let workflow = structured?["workflow"]
    metrics = (structured?["metrics"]?.arrayValue ?? []).map(LegacyMetricView.init)
    legacyStatus = workflow?["status"]?.stringValue
    completedCategories = (workflow?["completedCategories"]?.arrayValue ?? []).compactMap(\.stringValue)
  }

  public var trainerId: String? { stored["trainerId"]?.stringValue }
  public var memberId: String? { stored["memberId"]?.stringValue }

  /// Legacy `date` (epoch milliseconds, or a timestamp).
  public var sessionDate: Date? {
    switch stored["date"] {
    case let .int(millis)?: return Date(timeIntervalSince1970: Double(millis) / 1000)
    case let .timestamp(date)?: return date
    default: return nil
    }
  }

  /// v2 status for `legacyStatus` (V1-05 §5.6). Nil for a value outside the mapping.
  public var mappedStatus: SoapStatus? {
    switch legacyStatus {
    case nil, "draft"?, "작성 중"?: return .draft
    case "complete"?, "완료"?, "shared"?, "공유됨"?: return .finalized
    default: return nil
    }
  }

  /// The note was shared under v1 (status or flag). v2 does not share it again (MIG-05).
  public var wasShared: Bool {
    stored["isSharedWithMember"] == .bool(true) || legacyStatus == "shared" || legacyStatus == "공유됨"
  }

  /// `completedCategories` with `S·O·A·P` written as `subjective·objective·assessment·plan`.
  public var normalizedCompletedCategories: [String] {
    let names = ["S": "subjective", "O": "objective", "A": "assessment", "P": "plan"]
    return completedCategories.map { names[$0] ?? $0 }
  }

  /// False for an empty memberId or a local `member-` seed (MIG-07, '회원 연결 없음').
  public var hasLinkedMember: Bool {
    guard let id = memberId else { return false }
    return !id.isEmpty && !id.hasPrefix("member-")
  }

  /// Inline ink (`drawingData` bytes or `structured.subjective.nativeInkDataBase64`). Read only; v2 keeps
  /// ink in Storage (MIG-06).
  public var hasInlineInk: Bool {
    if case .bytes? = stored["drawingData"] { return true }
    let ink = stored["structured"]?["subjective"]?["nativeInkDataBase64"]?.stringValue
    return !(ink ?? "").isEmpty
  }
}

File: lib/models/legacy_soap_view.dart
// Read-only view of a v1 (legacy) SOAP document (DF-007, AC-SOAP-06.2,
// docs/v1/05_DATA_MODEL_AND_RULES.md §5.6, PRD §11.4).
//
// v1 documents have no `schemaVersion`. The view keeps every `structured.metrics` row and
// shows each one as 'uninterpretable' (해석 불가): no legacy row carries the condition fields
// a v2 metric needs (joint·motion·activeOrPassive, muscleGroup), so even a parsable number
// is not a catalog metric (V1-05 §5.6). Values are never filled with 0.
//
// The view does not expose `diagnosis` (MIG-04) and has no write path.

import 'package:cloud_firestore/cloud_firestore.dart' show Blob, Timestamp;
import 'package:flutter/foundation.dart' show immutable;

import '../contracts/generated/vocab.g.dart' show SoapStatus;

/// How a legacy metric row reads in v2.
enum LegacyMetricInterpretation {
  /// Kept as the original text and shown as '해석 불가' (F-SOAP-06.2).
  uninterpretable,
}

/// One `structured.metrics[]` row of a v1 document, kept as stored.
@immutable
final class LegacyMetricView {
  const LegacyMetricView(this.source);

  /// The stored element. Normally a map with `type, label, side, value, unit, score, note`.
  final Object? source;

  Map<String, Object?> get raw => source is Map
      ? Map<String, Object?>.unmodifiable({
          for (final e in (source as Map).entries) '${e.key}': e.value,
        })
      : const {};

  /// The legacy type as written: `rom`, `ROM`, `통증`, `test`, `general`, ...
  String? get type => _text(raw['type']);
  String? get label => _text(raw['label']);
  String? get side => _text(raw['side']);
  String? get unit => _text(raw['unit']);
  String? get note => _text(raw['note']);

  /// The value as stored (a string in every legacy writer), shown as text.
  String get valueText {
    final value = raw['value'];
    return value == null ? '' : '$value';
  }

  LegacyMetricInterpretation get interpretation =>
      LegacyMetricInterpretation.uninterpretable;

  bool get isInterpretable => false;

  /// The label shown next to the original text.
  String get displayLabel => LegacySoapView.uninterpretableLabel;
}

/// Read-only view of a v1 SOAP document.
@immutable
final class LegacySoapView {
  const LegacySoapView._({
    required this.raw,
    required this.metrics,
    required this.legacyStatus,
    required this.completedCategories,
  });

  /// Reads a v1 document (Firestore types; `$int` millis are ints).
  factory LegacySoapView.fromMap(Map<String, dynamic> data) {
    final raw = Map<String, Object?>.unmodifiable(data);
    final structured = _map(raw['structured']);
    final workflow = _map(structured?['workflow']);
    final rows = structured?['metrics'];
    final categories = workflow?['completedCategories'];
    return LegacySoapView._(
      raw: raw,
      metrics: List.unmodifiable([
        if (rows is List)
          for (final row in rows) LegacyMetricView(row),
      ]),
      legacyStatus: _text(workflow?['status']),
      completedCategories: List.unmodifiable([
        if (categories is List)
          for (final c in categories)
            if (c is String) c,
      ]),
    );
  }

  /// Display label for legacy rows and unknown v2 metric codes.
  static const String uninterpretableLabel = '해석 불가';

  /// The document as read.
  final Map<String, Object?> raw;

  /// Every `structured.metrics` row, in stored order. Never shorter than the stored list.
  final List<LegacyMetricView> metrics;

  /// `structured.workflow.status` as written (`complete`, `완료`, `작성 중`, `공유됨`, ...).
  final String? legacyStatus;

  /// `structured.workflow.completedCategories` as written (`S` or `subjective` style).
  final List<String> completedCategories;

  String? get trainerId => _text(raw['trainerId']);
  String? get memberId => _text(raw['memberId']);

  /// Legacy `date` (epoch millis) as UTC.
  DateTime? get sessionDate {
    final date = raw['date'];
    if (date is int) {
      return DateTime.fromMillisecondsSinceEpoch(date, isUtc: true);
    }
    if (date is Timestamp) return date.toDate().toUtc();
    return null;
  }

  /// v2 status for [legacyStatus] (V1-05 §5.6). Null for a value outside the mapping.
  SoapStatus? get mappedStatus => switch (legacyStatus) {
        null || 'draft' || '작성 중' => SoapStatus.draft,
        'complete' || '완료' || 'shared' || '공유됨' => SoapStatus.finalized,
        _ => null,
      };

  /// The note was shared under v1 (status or flag). v2 does not share it again (MIG-05).
  bool get wasShared =>
      raw['isSharedWithMember'] == true ||
      legacyStatus == 'shared' ||
      legacyStatus == '공유됨';

  /// [completedCategories] with `S·O·A·P` written as `subjective·objective·assessment·plan`.
  List<String> get normalizedCompletedCategories => [
        for (final c in completedCategories) _categoryNames[c] ?? c,
      ];

  /// False for an empty memberId or a local `member-` seed (MIG-07, '회원 연결 없음').
  bool get hasLinkedMember {
    final id = memberId;
    return id != null && id.isNotEmpty && !id.startsWith('member-');
  }

  /// Inline ink (`drawingData` bytes or `structured.subjective.nativeInkDataBase64`).
  /// Read only; v2 keeps ink in Storage (MIG-06).
  bool get hasInlineInk {
    if (raw['drawingData'] is Blob) return true;
    final subjective = _map(_map(raw['structured'])?['subjective']);
    final ink = subjective?['nativeInkDataBase64'];
    return ink is String && ink.isNotEmpty;
  }

  static const Map<String, String> _categoryNames = {
    'S': 'subjective',
    'O': 'objective',
    'A': 'assessment',
    'P': 'plan',
  };
}

String? _text(Object? value) => value is String ? value : null;

Map<Object?, Object?>? _map(Object? value) => value is Map ? value : null;
